### PR TITLE
Migrate CircleCI workflows to GitHub Actions (3/3)

### PR DESCRIPTION
**Pull request by AzfaarQureshi**
### What this PR does

This is PR 3/3 of migrating Cortex's CI from CircleCI to GitHub Actions: 

**[Part 1/3](https://github.com/cortexproject/cortex/pull/3302)** 
* create `test-build-deploy` workflow under `.github/workflows`
* add readme
* adding the 3 base jobs
   * `lint`
   * `test`
   * `build`
* these jobs are run async and share no dependencies with each other
* workflow is run when a commit is pushed to `master` or when any PR is opened. However CD jobs will only run on the master branch and will be skipped otherwise
* all jobs in following PRs are dependant on one or more of these jobs finishing

**[Part 2/3](https://github.com/cortexproject/cortex/pull/3341)** 
* adding jobs dependant on `build`
    * `integration`
    * `integration-config-db`
* This PR completes the CI portion of migration. The following PR will finish the remaining CD portion of the migration

**[Part 3/3](https://github.com/cortexproject/cortex/pull/3368)** 👈 
* This PR completes the migration from CircleCI to GitHub Actions
   * `deploy_website` is dependant on `build, test`
   * `deploy` is dependant on `build, test, lint, integration, integration-config-db`
* CD jobs will only run on the master branch and will be otherwise skipped

### Which issue(s) this PR fixes:
fixes https://github.com/cortexproject/cortex/issues/3274

### Checklist
 - [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pracucci** [comment]
@AzfaarQureshi I've merged the part 2 PR. Could you rebase `master`, please?

File: .github/workflows/test-build-deploy.yml [hunk 1]
```diff
@@ -75,15 +73,176 @@ jobs:
...
+    needs: [build, test]
+    if: github.ref == 'refs/heads/master' && github.repository == 'cortexproject/cortex'
+    runs-on: ubuntu-latest
+    container:
+      image: quay.io/cortexproject/build-image:upgrade-build-image-debian-491e60715-WIP
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v2
+        with:
+          ssh-key: ${{ secrets.SSH_PRIVATE_KEY }}
```
Comment (pracucci) [inline .github/workflows/test-build-deploy.yml:183]: Why is this required, contrary to other jobs?

File: .github/workflows/test-build-deploy.yml [hunk 2]
```diff
@@ -75,15 +73,176 @@ jobs:
...
+          ssh-agent -a $SSH_AUTH_SOCK > /dev/null
+          ssh-add - <<< "${{ secrets.SSH_PRIVATE_KEY }}"
+        env:
+          SSH_AUTH_SOCK: /tmp/ssh_agent.sock
+        shell: bash
+      - name: Deploy Website
+        run: make BUILD_IN_CONTAINER=false web-deploy
+        env:
+          SSH_AUTH_SOCK: /tmp/ssh_agent.sock
+          GIT_SSH_COMMAND: "ssh -o StrictHostKeyChecking=no"
```
Comment (pracucci) [inline .github/workflows/test-build-deploy.yml:206]: I would add a code comment to explain that the web-deploy script use git to checkout and push to the github pages repository and we use ssh to authenticate to github.

File: .github/workflows/test-build-deploy.yml [hunk 3]
```diff
@@ -75,15 +73,176 @@ jobs:
...
+      - name: Sym Link Expected Path to Workspace
+        run: |
+          mkdir -p /go/src/github.com/cortexproject/cortex
+          ln -s $GITHUB_WORKSPACE/* /go/src/github.com/cortexproject/cortex
+      - name: Download Website Artifact
+        uses: actions/download-artifact@v2
+        with:
+          name: website public
+          path: website/public
+      - name: Setup SSH Keys and known_hosts
```
Comment (pracucci) [inline .github/workflows/test-build-deploy.yml:193]:
```suggestion
      - name: Setup SSH Keys and known_hosts used to authenticated to GitHub to deploy website
```

File: .github/workflows/test-build-deploy.yml [hunk 4]
```diff
@@ -75,15 +73,176 @@ jobs:
...
+        run: |
+          mkdir -p ~/.ssh
+          ssh-keyscan github.com >> ~/.ssh/known_hosts
+          ssh-agent -a $SSH_AUTH_SOCK > /dev/null
+          ssh-add - <<< "${{ secrets.SSH_PRIVATE_KEY }}"
```
Comment (pracucci) [inline .github/workflows/test-build-deploy.yml:198]: May be worth renaming `SSH_PRIVATE_KEY` to `GITHUB_SSH_PRIVATE_KEY` for clarity?

File: .github/workflows/test-build-deploy.yml [hunk 5]
```diff
@@ -75,15 +73,176 @@ jobs:
...
+          fi
+          if [ -n "$QUAY_PASSWORD" ]; then
+            docker login -u "$QUAY_USER" -p "$QUAY_PASSWORD" quay.io;
+          fi
+          IMAGE_TAG=$GIT_TAG ./push-images $NOQUAY
+        env:
+          DOCKER_REGISTRY_PASSWORD: ${{ secrets.DOCKER_REGISTRY_PASSWORD }}
+          DOCKER_REGISTRY_USER: ${{secrets.DOCKER_REGISTRY_USER}}
+          QUAY_PASSWORD: ${{secrets.QUAY_PASSWORD}}
+          QUAY_USER: ${{secrets.QUAY_USER}}
```
Comment (pracucci) [inline .github/workflows/test-build-deploy.yml:246]: Similarly to `DOCKER_REGISTRY_*` I would call these env variables `QUAY_REGISTRY_PASSWORD` and `QUAY_REGISTRY_USER` for clarity.

**Review by pracucci — APPROVED** [review]
Thanks for working on it! I left few minor comments, but overall LGTM (good job!). Please remember to rebase `master`.

As soon as we merge this, the deploy and deploy_website jobs will both run in CircleCI and GitHub actions, effectively deploying the same thing twice. I would suggest to comment out (not remove for now) the deploy and deploy_website trigger from CircleCI so that we'll only run the GitHub actions once (and we'll soon realise if there's any issue, and in case we'll reiterate on it to fix it).

**Comment by pracucci** [comment]
Assuming you will pick my suggestions, I've already setup the following secrets:

- `DOCKER_REGISTRY_PASSWORD`
- `DOCKER_REGISTRY_USER`
- `QUAY_REGISTRY_PASSWORD`
- `QUAY_REGISTRY_USER`
- `GITHUB_SSH_PRIVATE_KEY`

I've intentionally not defined `NOQUAY` and `IMAGE_PREFIX` because the defaults should be fine.

File: .github/workflows/test-build-deploy.yml (hunk 2 above)
Comment (shovnik) [reply .github/workflows/test-build-deploy.yml:208]: Comment added

File: .github/workflows/test-build-deploy.yml (hunk 3 above)
Comment (shovnik) [reply .github/workflows/test-build-deploy.yml]: Amended

File: .github/workflows/test-build-deploy.yml (hunk 4 above)
Comment (shovnik) [reply .github/workflows/test-build-deploy.yml]: Renamed

File: .github/workflows/test-build-deploy.yml (hunk 5 above)
Comment (shovnik) [reply .github/workflows/test-build-deploy.yml]: Standardized registry env variable naming

**Comment by shovnik** [comment]
Rebased and made all suggested changes @pracucci . I commented out the deploy and deploy-website workflows in CircleCI to avoid deploying twice since they were using the default triggers with restrictions. Let us know if there is anything else we need to change. Thanks for all three reviews!

File: .github/workflows/test-build-deploy.yml (hunk 1 above)
Comment (AzfaarQureshi) [reply .github/workflows/test-build-deploy.yml]: I'll add a comment for further clarity but to answer your question:

The `web-deploy` script assumes the repo to be cloned with `.git` rather than `https` or else lines like [this](https://github.com/cortexproject/cortex/blob/master/tools/website/web-deploy.sh#L7) will return wrong urls. So the ssh key here is required so that web-deploy can correctly interact with github 😄

**Review by pracucci — APPROVED** [review]
LGTM, thanks! Pinging @pstibrany for the 2nd (required) review and then we can merge. In case of any issue, we can easily disable GitHub actions and de-comment CircleCI.

**Comment by shovnik** [comment]
Sounds good, thanks for the update.

File: .github/workflows/test-build-deploy.yml [hunk 6]
```diff
@@ -160,7 +165,87 @@ jobs:
...
+        run: |
+          ln -s /tmp/images ./docker-images
+          make BUILD_IN_CONTAINER=false load-images
+      - name: Deploy
+        run: |
+          if [ -n "$DOCKER_REGISTRY_PASSWORD" ]; then
+            docker login -u "$DOCKER_REGISTRY_USER" -p "$DOCKER_REGISTRY_PASSWORD"
+          fi
+          if [ -n "$QUAY_PASSWORD" ]; then
+            docker login -u "$QUAY_REGISTRY_USER" -p "$QUAY_REGISTRY_PASSWORD" quay.io;
```
Comment (pstibrany) [inline .github/workflows/test-build-deploy.yml:242]: Looks like extra semicolon at the end.

**Review by pstibrany — APPROVED** [review]
LGTM, thank you for all this work! (my comments/questions are not blocking)

File: .github/workflows/test-build-deploy.yml (hunk 6 above)
Comment (AzfaarQureshi) [reply .github/workflows/test-build-deploy.yml:242]: I included it here because it's in the circleci config as well: https://github.com/cortexproject/cortex/blob/master/.circleci/config.yml#L247

Do you want me to remove the semicolon in GHA?

File: .github/workflows/test-build-deploy.yml [hunk 7]
```diff
@@ -160,7 +165,87 @@ jobs:
...
+          ln -s $GITHUB_WORKSPACE/* /go/src/github.com/cortexproject/cortex
+      - name: Download Website Artifact
+        uses: actions/download-artifact@v2
+        with:
+          name: website public
+          path: website/public
+      - name: Setup SSH Keys and known_hosts for Github Authentication to Deploy Website
+        run: |
+          mkdir -p ~/.ssh
+          ssh-keyscan github.com >> ~/.ssh/known_hosts
```
[the remark this answers is not in the document]
Comment (AzfaarQureshi) [reply .github/workflows/test-build-deploy.yml:198]: Yeah I thought so too but removing this section causes a `could not resolve host: GitHub.com` 😞

File: .github/workflows/test-build-deploy.yml [hunk 8]
```diff
@@ -160,7 +165,87 @@ jobs:
...
+    needs: [build, test]
+    if: github.ref == 'refs/heads/master' && github.repository == 'cortexproject/cortex'
+    runs-on: ubuntu-latest
+    container:
+      image: quay.io/cortexproject/build-image:upgrade-build-image-debian-491e60715-WIP
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v2
+        with:
+          # web-deploy script requires repo to be cloned with ssh for some commands to work
```
[the remark this answers is not in the document]
Comment (AzfaarQureshi) [reply .github/workflows/test-build-deploy.yml:184]: The current CircleCI workflow also adds a ssh key because `git push` expects to communicate over ssh (because of how remote is set [here](https://github.com/cortexproject/cortex/blob/master/tools/website/web-deploy.sh#L7). You are right, we could use HTTPS with a token but that would require modifying `web-deploy.sh` to set the remote url with `https://cortexproject<TOKEN>@github.com/cortexproject/cortex.git` which would break CircleCI the way it is currently. 

As I type this out I realize `require` might be a strong word because we _could_ do it over https. Maybe rephrasing the comment as "web-deploy script _expects_ repo to be cloned with" would be better?

[the remark this answers is not in the document]
Comment (AzfaarQureshi) [reply .github/workflows/test-build-deploy.yml]: changed comment!

File: .github/workflows/test-build-deploy.yml (hunk 6 above)
Comment (pstibrany) [reply .github/workflows/test-build-deploy.yml:242]: Let's keep it for consistency for now.
